text/trace/0219-payload-collection.md

```diff
@@ -0,0 +1,341 @@
+<https://github.com/open-telemetry/oteps/pull/187>, which are very much related.
+A possible approach is to adopt changes based on this proposal for payload attributes only, hence not over-complicating
+mostly simple attributes.
+
+A further improvement is to add data classification to each `Value` object, to support the classification of
+internal values. For example, an `"email"` field inside a mapping could be specifically classified.
+
+Note that the separation itself of payload attributes from the rest is a logical classification to
+be leveraged by processors and backends to apply custom logic. For example, OTel collector could support
+dropping such attributes, and backends could use different encryption strategies.
+
+For instrumentations that collect payloads, we propose using a "safe by default" strategy, as proposed in
+<https://github.com/open-telemetry/oteps/pull/100> as well - payloads should support a "normalization"
+which is safe to assume that removes any sensitive data (such as values in SQL statements).
+When not possible, payloads will be collected only if explicitly configured by the user.
+
+## Prior art
+
+In Epsagon (now part of Cisco), we have already implemented payload collection in our previous (non-OTel) libraries.
+This capability was leveraged by many customers, including big companies that enabled it in production (and testing) environments.
+It was implemented in multiple runtimes (NodeJS, Python, Go, and others) and instrumentations (e.g. HTTP, AWS SDK, and many DB SDKs).
+
+In OpenTelemetry, there are existing requests for supporting payload collection -
+
+* <https://github.com/open-telemetry/opentelemetry-specification/issues/1062>
+* <https://github.com/open-telemetry/opentelemetry-specification/issues/376#issuecomment-1227501082>
+
+## Alternatives
+
+There are other possible ways to encode payload data, using current Span attributes.
+As we show, each has different drawbacks, and also missing the proposed metadata information (such as `original_encoding`, `encoded_size`
+and future fields as well).
+
+Also, explicitly separating payload attributes allows custom logic to be applied.
+This is especially true while currently there isn't a structured way to classify such attributes
+
+Potentially, we can update current Span attributes with all of the proposed functionality, though we argue
+that it will introduce over-complexity where it is mostly not necessary for simple attributes.
+
+### Encoding payload data as a serialized JSON attribute
+
+The major drawback of this alternative is that altering the data requires expensive deserialization and serialization.
+For example, this will be required by a simple processor that filters specific keys in a map.
+Also, it requires backends (and potentially processors) to unnecessarily attempt deserialization of every string attribute.
+
+### Supporting nested map values in Span attributes
```

Comment:
I think this alternate is not sufficiently explored. IMO, instead of adding a new set of attributes with a new type that mostly overlaps with AnyValue, this alternate can be more viable.

I would rather advocate that https://github.com/open-telemetry/opentelemetry-specification/issues/376 is accepted and come up with semantic convention to record the payload and associated metadata as a regular Span attribute.

This OTEP proposes and approach that requires much bigger changes, but I do not feel that the arguments in favour of it are strong enough.

Reply:
I agree that there is room to explore this further.

Some ideas (which I can also add to the OTEP):
* We can consider having a 'core' type (like AnyValue) which could be extended here
* Even if https://github.com/open-telemetry/opentelemetry-specification/issues/376 is approved, I think that this proposal introduces essential advantages - 
  * Adding the metadata as semantic conventions instead is quite cumbersome. We need 4 extra tags to include proposed metadata (`original_encoding`, `encoded_size`, `original_length` and `dropped_keys`). The last two should support nested values so that some keys will get duplicated. We may have more in the future.
  * Specifying attributes as payloads is important. For example, to disable all payload collection in Span SDK implementation, or in processors and exporters. Doing that by semantic conventions is less robust. And in general, naming tags like 'http.payload.request.body' isn't ideal.
* On the other hand,  there are some drawbacks to supporting nested maps in general attributes (as discussed in the other issue). So this proposal could be another alternative to that.

To sum up, I think that uniting both attribute types is possible, but we have to make sure they will be generic enough.
Doing that could over-complicate the most common use of 'simple' attributes and require even bigger changes than adding payload attributes.